### ISSUE TITLE
[4.x] Required field validation [a11y]

### ISSUE THREAD

**Issue by brianteeman** [joomla/joomla-cms#36642]
Whenever you have a required field a screen reader will always announce the field as _invalid_ and _required_ even though you havent done anything yet.

The correct solution for this is to add the attribute aria-invalid="false" when the field is first loaded and change that to aria-invalid="true" if the validation fails.

The problem is that the validate.js appears to be running onfocus and not onblur so the markInvalid function adds aria-invalid="true" to the field onfocus and every required field is announced as _required invalid_

There are two possible/related solutions

1. validate.js is change to onblur AND the attribute aria-invalid=false is added to the markup of all required fields
2. validate.js is changed to add aria-invalid-false onfocus and the markInvalid is changed to be triggered onblur

The easiest field to see this in action with is the admin login

[More Details](https://www.tpgi.com/required-attribute-requirements/)

cc @chmst @Fedik @dgrammatiko

**Comment by Fedik** [comment]
Or add `aria-invalid="false"` in a field layout, when field is set to `required`. And validate.js will change it to `aria-invalid="true"` on editing.

But not is that is desired behavior? to always announce the field as invalid and required.

**Comment by brianteeman** [comment]
it is not invalid until it has been validated

currently its telling you "you made a mistake" before you have even done anything

**Comment by Fedik** [comment]
Okay, then what about changing this https://github.com/joomla/joomla-cms/blob/50685f9147ed458be421e544591444e27ca4613b/layouts/joomla/form/field/text.php#L72

to:
```php
$required ? 'required aria-invalid="false"' : '',
```

On initial render it will produce:
```html
<input name="blabla" required aria-invalid="false" ...>
```

**Comment by dgrammatiko** [comment]
> `$required ? 'required aria-invalid="false"' : '',`

There's more than that here. The aria-invalid should be set **ONLY** if the value is null

**Comment by brianteeman** [comment]
And you have focused into the field and then blurred. Not just because it's empty on page load

**Comment by dgrammatiko** [comment]
The problem is actually in the validation.js or better IS the validation.js. The script is just a rewrite of some script with origins before 2010. There are more modern scripts that will solve better the task in hand. I have one for my own use cases: https://formally-valid.dgrammatiko.dev

**Comment by chmst** [comment]
@dgrammatiko thanks for this information. Do you see a chance to introduce you solution in Joomla?

**Comment by brianteeman** [comment]
While the script from @dgrammatiko is good it is not a drop in replacement.

**Comment by dgrammatiko** [comment]
@brianteeman is right, it's not a drop in replacement. Someone needs to check if some kind of fallback could be written for the old validators but still there are expected changes in the form XMLs (probably a default could be used and extended/overridden in the XML definition to make it b/c). The short version is it needs a bit of investigation how (or even IF) this could be done in a minor version. The other option is to plan it for v5 in 15 months or so...

**Comment by brianteeman** [comment]
I spent some time on this. As the script is not a drop in replacement it couldnt be done until J5 but I am not 100 convinced that it should be even then. There are other scripts that do not require changing the xml on all the files which would be a better fit. They might not be as complete but they are much less time to implement for core and extensions

**Comment by dgrammatiko** [comment]
> There are other scripts that do not require changing the xml on all the files which would be a better fit.

That's true but as I said if this is implemented with some default data-attr then not all fields would need entries. BTW the entries in the XML are the validation error messages, you can even skip the custom messages and let the browser provide whatever is the default per input. Most of the things are opt in, very few from all these attributes are the hard requirements, eg in the from element all these attributes are the default for Joomla/Bootstrap 5:
```html
<form 
data-valid-class="is-valid"
data-invalid-class="is-invalid"
data-invalid-form-alert="true"
data-indicator="true"
data-indicator-position="after"
data-indicator-class="invalid-feedback">
```

The attribute `data-indicator="true"` controls if the form should use the custom or the browser error messages. If this is `false`  then the input could be as simple as (same as is right now but there's no customised message :( )
```html
    <div class="col-md-4 mb-3">
      <label for="validationServerUsername">Username</label>
      <div class="input-group">
        <div class="input-group-prepend">
          <span class="input-group-text" id="inputGroupPrepend3">@</span>
        </div>
        <input type="text" class="form-control" id="validationServerUsername" aria-describedby="inputGroupPrepend3" required>
      </div>
    </div>
  </div>

``` 


There's a codepen [live example](https://codepen.io/dgrammatiko/pen/RwjbpRp?layout=top)

**Comment by brianteeman** [comment]
we really dont need to add the data-

**Comment by dgrammatiko** [comment]
> we really dont need to add the data-

This decision is not up to me but having a way to provide your own error messages is kinda good foe the UX, forms could be way more approachable (maybe not for the backend)...
